macOS x64 clang build failing

The `CI - macOS x64 clang` workflow started failing when scheduled for commit f1e1866 (build log https://github.com/iree-org/iree/actions/runs/12514308966) when linking `compiler/bindings/python/iree/compiler/_mlir_libs/_mlir.cpython-311-darwin.so`:

```
[7968/8273] Linking CXX shared module compiler/bindings/python/iree/compiler/_mlir_libs/_mlir.cpython-311-darwin.so
FAILED: compiler/bindings/python/iree/compiler/_mlir_libs/_mlir.cpython-311-darwin.so 
: && /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -O2 -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -mmacosx-version-min=13.7 -bundle -Wl,-headerpad_max_install_names -Wl,@/Users/runner/work/iree/iree/build-macos/_deps/nanobind-src/cmake/darwin-ld-cpython.sym -lm -o compiler/bindings/python/iree/compiler/_mlir_libs/_mlir.cpython-311-darwin.so compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/MainModule.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRAffine.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRAttributes.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRCore.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRInterfaces.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRModule.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRTypes.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/Pass.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/Rewrite.cpp.o  -Wl,-rpath,@loader_path -Wl,-rpath,/Users/runner/work/iree/iree/build-macos/lib  llvm-project/lib/libnanobind-static.a  lib/libIREECompiler.dylib  llvm-project/lib/libLLVMSupport.a  -lm  llvm-project/lib/libLLVMDemangle.a && :
ld: warning: ignoring duplicate libraries: '-lm'
ld: Undefined symbols:
  _PyClassMethod_New, referenced from:
      mlir::python::populateIRCore(nanobind::module_&) in IRCore.cpp.o
      nanobind::object classmethod<nanobind::object (*)(nanobind::object const&, std::__1::optional<nanobind::list>, nanobind::list, std::__1::optional<nanobind::dict>, std::__1::optional<std::__1::vector<mlir::python::PyBlock*, std::__1::allocator<mlir::python::PyBlock*>>>, std::__1::optional<int>, mlir::python::DefaultingPyLocation, nanobind::object const&), nanobind::arg, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, char const*>(nanobind::object (*)(nanobind::object const&, std::__1::optional<nanobind::list>, nanobind::list, std::__1::optional<nanobind::dict>, std::__1::optional<std::__1::vector<mlir::python::PyBlock*, std::__1::allocator<mlir::python::PyBlock*>>>, std::__1::optional<int>, mlir::python::DefaultingPyLocation, nanobind::object const&), nanobind::arg, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, char const*) in IRCore.cpp.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

The integrate commit f1e1866 updates LLVM to llvm/llvm-project@b135922. This includes llvm/llvm-project@b56d1ec with ports the Python core code to nanobind. The build only fails for macOS but passes for other platforms.

Till now, IREE's compiler bindings are build with pybind11. Porting IREE's compiler bindings from pybind11 to nanobind (PR #19580), the following (additional) targets fail for macOS which reference `_PyClassMethod_New`:
* `IREECompilerDialectsModule.cpp.o`
* `DialectGPU.cpp.o`
* `IRCore.cpp.o`
* `DialectPDL.cpp.o`
* `DialectLLVM.cpp.o`
* `DialectTransform.cpp.o`

Upstream macOS is several builds seem to pass, see for example https://github.com/makslevental/mlir-wheels/actions/runs/12565899455 or https://github.com/llvm/eudsl/actions/runs/12576273972/job/35052215749.

---
Errors (https://github.com/iree-org/iree/actions/runs/12583361482/job/35070746468) for PR #19580:  
```
[7969/8276] Linking CXX shared module compiler/bindings/python/iree/compiler/_mlir_libs/_ireeCompilerDialects.cpython-311-darwin.so
FAILED: compiler/bindings/python/iree/compiler/_mlir_libs/_ireeCompilerDialects.cpython-311-darwin.so 
: && /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -O2 -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -mmacosx-version-min=13.7 -bundle -Wl,-headerpad_max_install_names -Wl,@/Users/runner/work/iree/iree/build-macos/_deps/nanobind-src/cmake/darwin-ld-cpython.sym -lm -o compiler/bindings/python/iree/compiler/_mlir_libs/_ireeCompilerDialects.cpython-311-darwin.so compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._ireeCompilerDialects.dso.dir/IREECompilerDialectsModule.cpp.o  -Wl,-rpath,@loader_path -Wl,-rpath,/Users/runner/work/iree/iree/build-macos/lib  llvm-project/lib/libnanobind-static.a  lib/libIREECompiler.dylib  llvm-project/lib/libLLVMSupport.a  -lm  llvm-project/lib/libLLVMDemangle.a && :
ld: warning: ignoring duplicate libraries: '-lm'
ld: Undefined symbols:
  _PyClassMethod_New, referenced from:
      nanobind_init__ireeCompilerDialects(nanobind::module_&) in IREECompilerDialectsModule.cpp.o
      nanobind_init__ireeCompilerDialects(nanobind::module_&) in IREECompilerDialectsModule.cpp.o
      nanobind_init__ireeCompilerDialects(nanobind::module_&) in IREECompilerDialectsModule.cpp.o
      nanobind_init__ireeCompilerDialects(nanobind::module_&) in IREECompilerDialectsModule.cpp.o
      nanobind_init__ireeCompilerDialects(nanobind::module_&) in IREECompilerDialectsModule.cpp.o
      nanobind_init__ireeCompilerDialects(nanobind::module_&) in IREECompilerDialectsModule.cpp.o
      nanobind_init__ireeCompilerDialects(nanobind::module_&) in IREECompilerDialectsModule.cpp.o
      nanobind_init__ireeCompilerDialects(nanobind::module_&) in IREECompilerDialectsModule.cpp.o
      ...
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
```
[7971/8276] Linking CXX shared module compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsGPU.cpython-311-darwin.so
FAILED: compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsGPU.cpython-311-darwin.so 
: && /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -O2 -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -mmacosx-version-min=13.7 -bundle -Wl,-headerpad_max_install_names -Wl,@/Users/runner/work/iree/iree/build-macos/_deps/nanobind-src/cmake/darwin-ld-cpython.sym -lm -o compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsGPU.cpython-311-darwin.so compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlirDialectsGPU.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/DialectGPU.cpp.o  -Wl,-rpath,@loader_path -Wl,-rpath,/Users/runner/work/iree/iree/build-macos/lib  llvm-project/lib/libnanobind-static.a  lib/libIREECompiler.dylib  llvm-project/lib/libLLVMSupport.a  -lm  llvm-project/lib/libLLVMDemangle.a && :
ld: warning: ignoring duplicate libraries: '-lm'
ld: Undefined symbols:
  _PyClassMethod_New, referenced from:
      _PyInit__mlirDialectsGPU in DialectGPU.cpp.o
      _PyInit__mlirDialectsGPU in DialectGPU.cpp.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
```
[7973/8276] Linking CXX shared module compiler/bindings/python/iree/compiler/_mlir_libs/_mlir.cpython-311-darwin.so
FAILED: compiler/bindings/python/iree/compiler/_mlir_libs/_mlir.cpython-311-darwin.so 
: && /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -O2 -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -mmacosx-version-min=13.7 -bundle -Wl,-headerpad_max_install_names -Wl,@/Users/runner/work/iree/iree/build-macos/_deps/nanobind-src/cmake/darwin-ld-cpython.sym -lm -o compiler/bindings/python/iree/compiler/_mlir_libs/_mlir.cpython-311-darwin.so compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/MainModule.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRAffine.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRAttributes.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRCore.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRInterfaces.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRModule.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/IRTypes.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/Pass.cpp.o compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlir.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/Rewrite.cpp.o  -Wl,-rpath,@loader_path -Wl,-rpath,/Users/runner/work/iree/iree/build-macos/lib  llvm-project/lib/libnanobind-static.a  lib/libIREECompiler.dylib  llvm-project/lib/libLLVMSupport.a  -lm  llvm-project/lib/libLLVMDemangle.a && :
ld: warning: ignoring duplicate libraries: '-lm'
ld: Undefined symbols:
  _PyClassMethod_New, referenced from:
      mlir::python::populateIRCore(nanobind::module_&) in IRCore.cpp.o
      nanobind::object classmethod<nanobind::object (*)(nanobind::object const&, std::__1::optional<nanobind::list>, nanobind::list, std::__1::optional<nanobind::dict>, std::__1::optional<std::__1::vector<mlir::python::PyBlock*, std::__1::allocator<mlir::python::PyBlock*>>>, std::__1::optional<int>, mlir::python::DefaultingPyLocation, nanobind::object const&), nanobind::arg, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, char const*>(nanobind::object (*)(nanobind::object const&, std::__1::optional<nanobind::list>, nanobind::list, std::__1::optional<nanobind::dict>, std::__1::optional<std::__1::vector<mlir::python::PyBlock*, std::__1::allocator<mlir::python::PyBlock*>>>, std::__1::optional<int>, mlir::python::DefaultingPyLocation, nanobind::object const&), nanobind::arg, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, nanobind::arg_v, char const*) in IRCore.cpp.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
```
[7977/8276] Linking CXX shared module compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsLLVM.cpython-311-darwin.so
FAILED: compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsLLVM.cpython-311-darwin.so 
: && /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -O2 -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -mmacosx-version-min=13.7 -bundle -Wl,-headerpad_max_install_names -Wl,@/Users/runner/work/iree/iree/build-macos/_deps/nanobind-src/cmake/darwin-ld-cpython.sym -lm -o compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsLLVM.cpython-311-darwin.so compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlirDialectsLLVM.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/DialectLLVM.cpp.o  -Wl,-rpath,@loader_path -Wl,-rpath,/Users/runner/work/iree/iree/build-macos/lib  llvm-project/lib/libnanobind-static.a  lib/libIREECompiler.dylib  llvm-project/lib/libLLVMSupport.a  -lm  llvm-project/lib/libLLVMDemangle.a && :
ld: warning: ignoring duplicate libraries: '-lm'
ld: Undefined symbols:
  _PyClassMethod_New, referenced from:
      populateDialectLLVMSubmodule(nanobind::module_ const&) in DialectLLVM.cpp.o
      populateDialectLLVMSubmodule(nanobind::module_ const&) in DialectLLVM.cpp.o
      populateDialectLLVMSubmodule(nanobind::module_ const&) in DialectLLVM.cpp.o
      populateDialectLLVMSubmodule(nanobind::module_ const&) in DialectLLVM.cpp.o
      populateDialectLLVMSubmodule(nanobind::module_ const&) in DialectLLVM.cpp.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
```
[7980/8276] Linking CXX shared module compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsPDL.cpython-311-darwin.so
FAILED: compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsPDL.cpython-311-darwin.so 
: && /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -O2 -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -mmacosx-version-min=13.7 -bundle -Wl,-headerpad_max_install_names -Wl,@/Users/runner/work/iree/iree/build-macos/_deps/nanobind-src/cmake/darwin-ld-cpython.sym -lm -o compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsPDL.cpython-311-darwin.so compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlirDialectsPDL.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/DialectPDL.cpp.o  -Wl,-rpath,@loader_path -Wl,-rpath,/Users/runner/work/iree/iree/build-macos/lib  llvm-project/lib/libnanobind-static.a  lib/libIREECompiler.dylib  llvm-project/lib/libLLVMSupport.a  -lm  llvm-project/lib/libLLVMDemangle.a && :
ld: warning: ignoring duplicate libraries: '-lm'
ld: Undefined symbols:
  _PyClassMethod_New, referenced from:
      populateDialectPDLSubmodule(nanobind::module_ const&) in DialectPDL.cpp.o
      populateDialectPDLSubmodule(nanobind::module_ const&) in DialectPDL.cpp.o
      populateDialectPDLSubmodule(nanobind::module_ const&) in DialectPDL.cpp.o
      populateDialectPDLSubmodule(nanobind::module_ const&) in DialectPDL.cpp.o
      populateDialectPDLSubmodule(nanobind::module_ const&) in DialectPDL.cpp.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
```
[7983/8276] Linking CXX shared module compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsTransform.cpython-311-darwin.so
FAILED: compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsTransform.cpython-311-darwin.so 
: && /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -O2 -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -mmacosx-version-min=13.7 -bundle -Wl,-headerpad_max_install_names -Wl,@/Users/runner/work/iree/iree/build-macos/_deps/nanobind-src/cmake/darwin-ld-cpython.sym -lm -o compiler/bindings/python/iree/compiler/_mlir_libs/_mlirDialectsTransform.cpython-311-darwin.so compiler/bindings/python/CMakeFiles/IREECompilerPythonModules.extension._mlirDialectsTransform.dso.dir/__/__/__/third_party/llvm-project/mlir/lib/Bindings/Python/DialectTransform.cpp.o  -Wl,-rpath,@loader_path -Wl,-rpath,/Users/runner/work/iree/iree/build-macos/lib  llvm-project/lib/libnanobind-static.a  lib/libIREECompiler.dylib  llvm-project/lib/libLLVMSupport.a  -lm  llvm-project/lib/libLLVMDemangle.a && :
ld: warning: ignoring duplicate libraries: '-lm'
ld: Undefined symbols:
  _PyClassMethod_New, referenced from:
      populateDialectTransformSubmodule(nanobind::module_ const&) in DialectTransform.cpp.o
      populateDialectTransformSubmodule(nanobind::module_ const&) in DialectTransform.cpp.o
      populateDialectTransformSubmodule(nanobind::module_ const&) in DialectTransform.cpp.o
      populateDialectTransformSubmodule(nanobind::module_ const&) in DialectTransform.cpp.o
      populateDialectTransformSubmodule(nanobind::module_ const&) in DialectTransform.cpp.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Just a heads up (because I've been pulling my hair out for hours somewhere else about  this; `-Wl,-U` **is not correct for linux**:

`-DCMAKE_MODULE_LINKER_FLAGS="-Wl,-U -Wl,_PyClassMethod_New"`

and what will happen is you'll get the weirdest 

```
note: /usr/bin/ld: -r and -shared may not be used together
```

because the correct flag on linux is `-Wl,-u`. Quite infuriatingly the `u/U` flags mean basically opposite (or something) things on the two platforms:

Infuriatingly they mean the exact opposite things on the platforms:

# mac

## lld

```
-u symbol_name
             Specified that symbol symbol_name must be defined for the link to succeed.  
             This is useful to force selected functions to be loaded from a static library.

-U symbol_name
             Specified that it is ok for symbol_name to have no definition.  
             With -two_levelnamespace, the resulting symbol will be marked dynamic_lookup which means dyld will search all loaded images.
```

# linux 

## lld

```
  --undefined-glob=<pattern>
                          Force undefined symbol during linking
  --undefined-version     Allow unused version in version script (disabled by default)
  --undefined=<symbol>    Force undefined symbol during linking
  -u <value>              Alias for --undefined
```

## ld

```
  -Ur                         Build global constructor/destructor tables
  -u SYMBOL, --undefined SYMBOL
                              Start with undefined reference to SYMBOL
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -649,11 +649,6 @@ if(IREE_BUILD_PYTHON_BINDINGS)
   # can be troublesome. Note that nanobind requires FindPython.
   set(Python_EXECUTABLE "${Python3_EXECUTABLE}")
   find_package(Python 3.9 COMPONENTS Interpreter Development.Module NumPy REQUIRED)
-  if(APPLE)
-    # Workaround for upstream using PyClassMethod_New but nanobind not being aware
-    # see https://github.com/iree-org/iree/issues/19591
-    set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} -Wl,-U -Wl,_PyClassMethod_New")
-  endif()
 elseif(IREE_BUILD_COMPILER OR IREE_BUILD_TESTS)
   find_package(Python3 COMPONENTS Interpreter REQUIRED)
   set(Python_EXECUTABLE "${Python3_EXECUTABLE}")

